File: code/level2_ttest_parc.R
#!/usr/bin/env Rscript
# load funcs variables
project_dir <- system("source ~/.bash_profile; echo ${data_dir}", intern=T)
code_dir = paste(project_dir, "code", sep="/")
source(paste(code_dir,"funcs_get_vars.R", sep='/'))

# load packages
library(plyr)
library(dplyr)
library(tidyr)
if(!require(pracma)) install.packages("pracma"); require(pracma)
if(!require(oro.nifti)) install.packages("oro.nifti"); require(oro.nifti)

# MNI parcellation
parcellation <- round(readNIfTI(MNI_PARCELLATION))

corr_label='spear'
val_label=if (corr_label=='spear') 'r' else 'beta'

out_fname <- paste0("stat-",STAT, 
                    "_corr-", corr_label,
                    "_parc-", PARC_LAB,
                    "_tail-1")


# function to save stats into nifti
save_parcel <- function(values,file){
  outmap <- parcellation
  outmap@datatype <- 16
  outmap@bitpix <- 32
  for (i in 1:N_PARCELS){
    outmap@.Data[outmap@.Data==i] <- values[i]
  }
  writeNIfTI(outmap,file)
  print(paste(file, "saved."))
}

# read in all subject files into single data frame
rsa_dir <- paste0(RSA_DIR,'/*/',corr_label)
out_dir <- paste0(SECOND_LEVEL_DIR,'/parc',PARC_LAB,'/')
data_fnames <- dir(Sys.glob(rsa_dir),
                   pattern = paste0("\\_stat-", STAT,
                                    "_corr-", corr_label,
                                    "_parc-", PARC_LAB,
                                    "_val-", val_label,
                                    "_roi_stats.csv$"),
                   recursive=F, full.names = T)
print('Reading in files: ')
print(data_fnames)
all_data <- do.call(rbind,lapply(data_fnames,read.csv))
all_data <- as.data.frame(all_data)
# drop row numbers
all_data <- all_data[,!(names(all_data) %in% c('X'))]
orig_data <- all_data

# get predictors
predictors <- unique(all_data$predictor)
print("Predictors found: ")
print(predictors)

# get tasks
tasks <- unique(all_data$task)
print("Tasks found: ")
print(tasks)

# get list of ROIs
rois <- unique(all_data$roi)
print("Number of ROIs found: ")
print(length(rois))
if (length(rois) != N_PARCELS){
  print("WARNING: number of ROIs does not match N_PARCELS:")
  print(str(length(rois)), PARC_LAB)
}

# separate tasks into 2 columns
all_data_task <- all_data %>% spread(key = task, value = r)

# # calculate average across tasks
# all_data_task$avg <- (all_data_task$friend + all_data_task$number)/2

# calculate differences
print("Calculating differences")
all_data_task$diff_rel <- NA
all_data_task[all_data_task$predictor=='dist_cat-sn','diff_rel'] <- all_data_task[all_data_task$predictor=='dist_cat-sn','friend'] - all_data_task[all_data_task$predictor=='dist_cat-sn','number']
all_data_task[all_data_task$predictor=='deg_cat-sn','diff_rel'] <- all_data_task[all_data_task$predictor=='deg_cat-sn','number'] - all_data_task[all_data_task$predictor=='deg_cat-sn','friend']

# get significance levels and corrected p-values
get_sig <- function(p.value, sig_levels=c(0.001, 0.01, 0.05, 0.10),
                    sig_symbols = c('***','**','*','+'), ns_symbol='n.s.',
                    bonf_correction=T,
                    n_tests = 1)
  {
    # order significance levels smallest to largest
    sig_levels <- sort(sig_levels)
    sig_symbols <- sig_symbols[order(sig_levels)]
    # correct significance levels for number of tests
    if (bonf_correction) {
      sig_levels <- sig_levels/n_tests
      corr_p.value <- p.value * n_tests
    }
    # iterate through levels and return relevant symbol
    sig_symbol <- ns_symbol
    for (i in seq(length(sig_levels))){
      if (p.value < sig_levels[i]) {
        sig_symbol <- sig_symbols[i]
        break
      }
    }
    return(c(sig_symbol, corr_p.value))
  }

ttest_output <- function(data_vector, n_tests1 = 1, n_tests2 = 1, 
                         alternative="greater", 
                         mu = 0, conf.level = 0.95)
{
  tres <- t.test(data_vector, alternative="greater", mu = 0, conf.level = 0.95)
  pval <- tres$p.value[[1]]
  rval <- tres$estimate[[1]] # mean
  tval <- tres$statistic[[1]]
  dfs <- tres$parameter[[1]]
  #sig_pcor1 <- get_sig(pval, n_tests=n_tests1)
  #pval_corr1 <- as.numeric(sig_pcor1[2])
  #sig_symb1 <- sig_pcor1[1]
  #sig_pcor2 <- get_sig(pval, n_tests=n_tests2)
  #pval_corr2 <- as.numeric(sig_pcor2[2])
  #sig_symb2 <- sig_pcor2[1]
  # figure out which direction difference is going in
  #res <- c(rval, tval, dfs, pval, pval_corr1, sig_symb1, pval_corr2, sig_symb2)
  res <- c(rval, tval, dfs, pval)
  return(res)
}


# iterate through each ROI and perform t-test
print("Running one-sample, one-tailed t-tests")
results_diffrel <- c()
cnames_diffrel <- c('roi', 'predictor', 'r', 't', 'df', 'p_uncorr' ) 
                    #, 'p_corr_all','sig_all', 'p_corr_gm', 'sig_gm')
results_avg <- c()
cnames_avg <- c('roi', 'predictor', 'r', 't', 'df', 'p_uncorr' ) 
for (r in unique(all_data_task$roi)) {
  # subselect this ROI's data
  roi_df <- filter(all_data_task, roi==r)
  # run t-test on each predictor
  for (p in predictors){
    df <- filter(roi_df, predictor==p)
    # test relavent difference
    if (p %in% c("dist_cat-sn", "deg_cat-sn")){
      res <- ttest_output(df$diff_rel, n_tests1 = length(rois), n_tests2 = length(gm_rois))
      results_diffrel <- rbind(results_diffrel, c(r, p, res))
    }
    # test average r
    res <- ttest_output(df$avg)
    results_avg <- rbind(results_avg, c(r, p, res))
  }
}
# convert to data frames
results_diffrel <- as.data.frame(results_diffrel)
colnames(results_diffrel) <- cnames_diffrel
results_diffrel$p_uncorr <- as.numeric(as.character(results_diffrel$p_uncorr))
results_avg <- as.data.frame(results_avg)
colnames(results_avg) <- cnames_avg
results_avg$p_uncorr <- as.numeric(as.character(results_avg$p_uncorr))

# calculate corrected p-values
for (pred in c("dist_cat-sn", "deg_cat-sn")){#predictors){
  pvals_diffrel <- results_diffrel[results_diffrel$predictor == pred,"p_uncorr"]
  pvals_avg <- results_avg[results_avg$predictor == pred, 'p_uncorr']
  for (pam in p.adjust.methods){
    if (sum(results_diffrel$predictor == pred) > 0){
      pvals_diffrel_adjust <- p.adjust(pvals_diffrel, method=pam)
      results_diffrel[results_diffrel$predictor == pred,pam] <- pvals_diffrel_adjust
      # only save nifti if there is a significant parcel
      if (sum(pvals_diffrel_adjust < .05) > 0){
        fname <- paste0(out_dir,
                        out_fname,
                        "_pred-", pred,
                        "_effect-relevant",
                        "_pcorr-",pam,
                        "_dir-rev")
        save_parcel(1-pvals_diffrel_adjust, fname)
      }
    }
    
    pvals_avg_adjust <- p.adjust(pvals_avg, method=pam)
    results_avg[results_avg$predictor == pred,pam] <- pvals_avg_adjust
    # only save nifti if there is a significant parcel
    if (sum(pvals_avg_adjust < .05) > 0){
      fname <- paste0(out_dir,
                      out_fname,
                      "_pred-", pred,
                      "_effect-main",
                      "_pcorr-",pam,
                      "_dir-rev")
      save_parcel(1-pvals_avg_adjust, fname)
    }
  }
}

# write out csvs
fname = paste0(out_dir, 
               out_fname,
               "_effect-relevant.csv")
write.csv(results_diffrel, row.names=F,
          file = fname)
print(paste("File saved:", fname))

fname = paste0(out_dir, 
               out_fname,
               "_effect-main.csv")
write.csv(results_avg, row.names=F,
          file = fname)
print(paste("File saved:", fname))

File: code/level2_ttest_parc.py
#!/bin/python3

import os
import shutil
import sys
from datetime import datetime
from copy import deepcopy
import glob
import pandas as pd
import nibabel as nib
import numpy as np
from scipy.stats import ttest_rel, ttest_ind, ttest_1samp
from funcs import *

def load_nii( imgpath ):
    img4D = nib.load(imgpath, mmap=False).get_data()
    return img4D

def save_nii( data, refnii, filename ):
    out_img = nib.Nifti1Image(data, refnii.affine, refnii.header)
    out_img.to_filename(filename)
    print(str(datetime.now()) + ": File %s saved." % filename)

def save_parcel( parcel_vals, filename, parcellation=MNI_PARCELLATION ):
    parcel_vals = np.array(parcel_vals)
    # MNI parcellationlation
    parcellation = nib.load(MNI_PARCELLATION, mmap=False)
    parcellation.set_data_dtype(np.float)
    p_data = parcellation.get_data()
    out_data = deepcopy(p_data) * 1.0
    for i in range(N_PARCELS):
        p = (i+1)*1.0
        out_val = 1-parcel_vals[i]
        out_data[ p_data==p ] = out_val
    save_nii(out_data, parcellation, filename)

def correct_p (t,p):
    if t < 0:
        # if negative t-value, then relevant context was lower than non-relavant
        p = 1.
    else:
        p = p/2
    return(t,p)

def FDR( x, n=None ):
    """
    Assumes a list or numpy array x which contains p-values for multiple tests
    Copied from p.adjust function from R
    """
    if n is None:
        n = len(x)
    o = [ i[0] for i in sorted(enumerate(x), key=lambda v:v[1],reverse=True) ]
    ro = [ i[0] for i in sorted(enumerate(o), key=lambda v:v[1]) ]
    q = sum([1.0/i for i in range(1,n+1)])
    l = [ q*len(x)/i*x[j] for i,j in zip(reversed(range(1,n+1)),o) ]
    l = [ l[k] if l[k] < 1.0 else 1.0 for k in ro ]
    return l

parc_label = str(N_PARCELS)

if len(sys.argv) > 1:
    corr_labels = sys.argv[1:]
else:
    corr_labels = ['spear', 'reg']

in_dir = RSA_DIR
out_dir = SECOND_LEVEL_DIR

for corr_label in corr_labels:
    print('Starting analysis of '+corr_label+' csv files')
    # read in all subject files into single data frame
    out_fname = "stat-"+STAT+"_corr-"+ corr_label+"_parc-"+ parc_label+"_tail-1"
    fnames = os.path.join(in_dir, "sub-*", corr_label, "parc",
                             "*_stat-"+ STAT+
                             "_corr-"+ corr_label+
                             "_parc-"+ parc_label+
                             "_roi_stats.csv")
    data_fnames = glob.glob(fnames)

    print('Reading in files: ')
    print(data_fnames)
    if len(data_fnames)==0:
          print('no files found matching: '+fnames)
          continue
    all_data = pd.concat((pd.read_csv(f) for f in data_fnames))
    # drop row numbers
    all_data = all_data.iloc[:,1:]
    orig_data = deepcopy(all_data)

    # get predictors
    predictors = pd.unique(all_data['predictor'])
    print("Predictors found: ")
    print(predictors)

    if 'task' in all_data.columns:
        has_task=True
        # get tasks
        tasks = pd.unique(all_data['task'])
        print("Tasks found: ")
        print(tasks)
    else:
        has_task=False

    # get list of ROIs
    rois = pd.unique(all_data['roi'])
    print("Number of ROIs found: ")
    print(len(rois))
    if len(rois) != N_PARCELS:
      print("WARNING: number of ROIs does not match N_PARCELS:")
      print(str(len(rois)), parc_label)

    if has_task:
        # separate tasks into 2 columns
        all_data = all_data.pivot_table(index=['sub','roi','predictor'],
                                             columns='task', values='r')
        # reset header
        all_data.reset_index(inplace=True)
        all_data.columns.name = ''
        # calculate average across tasks
        all_data['avg'] = (all_data['friend'] + all_data['number'])/2
    else:
        # for regression output, save beta values under "avg" heading
        all_data['avg'] = all_data['beta']

    pred_dfs = []
    diff_dfs = []
    avg_dfs = []
    cnames = ['corr','test','estimate','pred','roi','t','df','p']
    for pred in predictors:
        print('Running one-way t-tests for predictor: '+pred)
        # subselect predictor's rows
        pred_df = all_data[all_data['predictor']==pred]
        # initialize out dataframes
        diff_df = pd.DataFrame(None, columns=cnames)
        avg_df = pd.DataFrame(None, columns=cnames)
        # iterate through parcels
        for i,r in enumerate(rois):
            print("Testing parcel "+ str(r)+ " ("+str(i)+"/"+str(len(rois)) +")")
            df = pred_df[pred_df['roi']==r]
            # task-specific tests
            if has_task:
                if pred=='deg':
                    rel_col = 'number'
                    nonrel_col = 'friend'
                elif pred=='dist':
                    rel_col = 'friend'
                    nonrel_col = 'number'
                # run two-sample t-test
                t,p = ttest_rel(df[rel_col], df[nonrel_col])
                # correct to one-sample
                t,p = correct_p(t,p)
                # degrees of freedom
                t_df = len(df[rel_col]) - 1
                diff = np.mean(df[rel_col] - df[nonrel_col])
                diff_df.loc[r]=[corr_label,'diff',diff,pred,r,t,t_df,p]
            # test where the predictor is encoded across tasks
            t,p = ttest_1samp(df['avg'], popmean = 0.)
            # correct p-value to one-sample
            t,p = correct_p(t,p)
            # degrees of freedom
            t_df = len(df['avg']) - 1
            estimate=np.mean(df['avg'])
            avg_df.loc[r]=[corr_label,'avg',estimate,pred,r,t,t_df,p]

        # correct full dataframe for multiple comparisons
        if has_task:
            diff_df['FDR'] = FDR(np.array(diff_df['p']))
        avg_df['FDR'] = FDR(np.array(avg_df['p']))

        # save output
        fname = out_dir+'parc-'+parc_label+'_pred-'+pred+'_test-%s_stat-'+STAT+'_corr-'+corr_label
        fname_csv = fname+'.csv'
        fname_nii = fname+'_p-%s'
        if has_task:
            diff_df.to_csv(fname_csv % ('diff'))
            save_parcel(diff_df['FDR'], fname_nii % ('diff','FDR'))
            save_parcel(diff_df['p'], fname_nii % ('diff','uncorrected'))
        avg_df.to_csv(fname_csv % ('avg'))
        save_parcel(avg_df['FDR'], fname_nii % ('avg', 'FDR'))
        save_parcel(avg_df['p'], fname_nii % ('avg','uncorrected'))
